graylog2-server/src/main/java/org/graylog2/inputs/transports/KafkaTransport.java

```diff
@@ -173,17 +197,158 @@ public void setMessageAggregator(CodecAggregator ignored) {
+                if (paused) {
+                    // we try not to spin here, so we wait until the lifecycle goes back to running.
+                    LOG.debug("Message processing is paused, blocking until message processing is turned back on.");
+                    Uninterruptibles.awaitUninterruptibly(pausedLatch);
+                }
+                // check for being stopped before actually getting the message, otherwise we could end up losing that message
+                if (stopped) {
+                    break;
+                }
+                if (isThrottled()) {
+                    blockUntilUnthrottled();
+                }
+
+                // process the message, this will immediately mark the message as having been processed. this gets tricky
+                // if we get an exception about processing it down below.
+                final byte[] bytes = consumerIterator.next().value();
+
+                // it is possible that the message is null
+                if (bytes == null) {
+                    continue;
+                }
+                totalBytesRead.addAndGet(bytes.length);
+                lastSecBytesReadTmp.addAndGet(bytes.length);
+
+                final RawMessage rawMessage = new RawMessage(bytes);
+                input.processRawMessage(rawMessage);
+            }
+        }
+
+        private Optional<Iterator<ConsumerRecord<byte[], byte[]>>> tryPoll() {
+            try {
+                // Workaround https://issues.apache.org/jira/browse/KAFKA-4189 by calling wakeup()
+                final ScheduledFuture<?> future = scheduler.schedule(consumer::wakeup, 2000, TimeUnit.MILLISECONDS);
+                final ConsumerRecords<byte[], byte[]> consumerRecords = consumer.poll(1000);
+                future.cancel(true);
+
+                return Optional.of(consumerRecords.iterator());
+            } catch (WakeupException e) {
+                LOG.error("WakeupException in poll. Kafka server is not responding.");
+            } catch (InvalidOffsetException | AuthorizationException e) {
+                LOG.error("Exception in poll.", e);
+            }
+            return Optional.empty();
+        }
+
+        @Override
+        public void run() {
+            while (!stopped) {
+                final Optional<Iterator<ConsumerRecord<byte[], byte[]>>> consumerIterator;
+                try {
+                    consumerIterator = tryPoll();
+                    if (! consumerIterator.isPresent()) {
+                        LOG.error("Caught recoverable exception. Retrying");
+                        Thread.sleep(2000);
+                        continue;
+                    }
+                } catch (KafkaException | InterruptedException e) {
+                    LOG.error("Caught unrecoverable exception in poll. Restarting input", e);
+                    // (Ab)use serverEventBus to properly restart the entire input.
+                    serverEventBus.post(InputCreated.create(input.getId()));
```

Comment:
This is ugly. Maybe we should just not do this at all.
I might be misinterpreting the `KafkaException` being not retryable.
Furthermore, I don't know how to trigger one, so this isn't a "real" problem.